packages/gitgraph-core/src/commit.ts

```diff
@@ -0,0 +1,174 @@
+import { CommitStyle } from "./template";
+import Branch from "./branch";
+
+export interface CommitOptions {
+  author: string;
+  subject: string;
+  style: CommitStyle;
+  body?: string;
+  notes?: string;
+  refs?: string[];
+  tree?: string;
+  hash?: string;
+  parents?: string[];
+  innerText?: string;
+}
+
+/**
+ * Generate a random hash.
+ *
+ * @return {string} hex string with 40 chars
+ */
+const getRandomHash = () =>
+  (
+    Math.random()
+      .toString(16)
+      .substring(3) +
+    Math.random()
+      .toString(16)
+      .substring(3) +
+    Math.random()
+      .toString(16)
+      .substring(3) +
+    Math.random()
+      .toString(16)
+      .substring(3)
+  ).substring(0, 40);
+
+export class Commit {
+  /**
+   * Ref names (injected by Gitgraph.withRefsAndTags)
+   */
+  public refs: Array<Branch["name"] | "HEAD"> = [];
+  /**
+   * Commit x position (injected by Gitgraph.withPosition)
+   */
+  public x: number = 0;
+  /**
+   * Commit y position (injected by Gitgraph.withPosition)
+   */
+  public y: number = 0;
+  /**
+   * Commit hash
+   */
+  public hash: string;
+  /**
+   * Abbreviated commit hash
+   */
+  public hashAbbrev: string;
+  /**
+   * Tree hash
+   */
+  public tree: string;
+  /**
+   * Abbreviated tree hash
+   */
+  public treeAbbrev: string;
+  /**
+   * Parent hashes
+   */
+  public parents: Array<Commit["hash"]>;
+  /**
+   * Abbreviated parent hashed
+   */
+  public parentsAbbrev: Array<Commit["hashAbbrev"]>;
+  /**
+   * Author
+   */
+  public author: {
+    /**
+     * Author name
+     */
+    name: string;
+    /**
+     * Author email
+     */
+    email: string;
+    /**
+     * Author date
+     */
+    timestamp: number;
+  };
+  /**
+   * Committer
+   */
+  public committer: {
+    /**
+     * Commiter name
+     */
+    name: string;
+    /**
+     * Commiter email
+     */
+    email: string;
+    /**
+     * Commiter date
+     */
+    timestamp: number;
+  };
+  /**
+   * Subject
+   */
+  public subject: string;
+  /**
+   * Body
+   */
+  public body: string;
+  /**
+   * Notes
+   */
+  public notes: string;
+  /**
+   * Style
+   */
+  public style: CommitStyle;
+  /**
+   * Inner text
+   */
+  public innerText?: string;
+  /**
+   * List of branches attached (injected by Gitgraph.withBranches)
+   */
+  public branches?: Array<Branch["name"]>;
```

Comment:
I wonder if that wouldn't be better not to be optional. Having to deal with undefined value is a bit painful. 

Plus, `Gitgraph.withBranches()` injects it so technically, at runtime, when we get rendered data this is never `undefined`.

We could default it to an empty array. What do you think?

P.S. same question for `tags`.

Reply:
I'm totally agree with this, the main challenge is to find a good name for commit in each state (eg: `RawCommit` `CommitWithBranch`…).

Ideally I want to have strong type on all the process, without using `as` or `!` to hack typescript ;)